### ISSUE TITLE
"Callback was already called" exception when attempting to pay invoice

### ISSUE THREAD

**Issue by Kixunil** [apotdevin/thunderhub#158]
**Describe the problem/bug**
Paying an invoice from ThunderHub failes with an exception "Callback was already called".

**Your environment**
* Version of ThunderHub: 0.9.16
* Deployment method: Debian package
* Other relevant environment details: nginx with special configuration (see below)

**To Reproduce**
1. Create an invoice in another node
2. Click Send in ThunderHub
3. Paste the invoice
4. Click send

**Expected behavior**
The invoice is paid (or a message about path not being found is shown, however in this case I tried with a directly-connected node)

**Actual behavior**
Red empty box appears followed by another box saying "unable to connect to node".
Exception in the log.

<details>
<summary>Log contents</summary>

```
Oct 25 17:59:55 polisnode thunderhub[10086]: 2020-10-25 17:59:55 error [THUB]: [ 400,
Oct 25 17:59:55 polisnode thunderhub[10086]:   'ExpectedAuthenticatedLndToRetrieveChainBalance',
Oct 25 17:59:55 polisnode thunderhub[10086]:   [length]: 2 ]
Oct 25 17:59:55 polisnode thunderhub[10086]: 2020-10-25 17:59:55 error [THUB]: [ 400, 'ExpectedLndForPendingChainBalance', [length]: 2 ]
Oct 25 17:59:55 polisnode thunderhub[10086]: 2020-10-25 17:59:55 error [THUB]: [ 400, 'ExpectedLndGrpcApiForChannelBalanceQuery', [length]: 2 ]
Oct 25 18:05:21 polisnode thunderhub[10086]: 2020-10-25 18:05:21 warn [THUB]: No SSO account available
Oct 25 18:08:34 polisnode thunderhub[10086]: /usr/lib/thunderhub/node_modules/async/internal/onlyOnce.js:9
Oct 25 18:08:34 polisnode thunderhub[10086]:         if (fn === null) throw new Error("Callback was already called.");
Oct 25 18:08:34 polisnode thunderhub[10086]:                          ^
Oct 25 18:08:34 polisnode thunderhub[10086]: Error: Callback was already called.
Oct 25 18:08:34 polisnode thunderhub[10086]:     at /usr/lib/thunderhub/node_modules/async/internal/onlyOnce.js:9:32
Oct 25 18:08:34 polisnode thunderhub[10086]:     at args.lnd.(anonymous function).(anonymous function) (/usr/lib/thunderhub/node_modules/ln-service/node_modules/lightning/lnd_methods/info/get_node.js:101:20)
Oct 25 18:08:34 polisnode thunderhub[10086]:     at Object.onReceiveStatus (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:1212:9)
Oct 25 18:08:34 polisnode thunderhub[10086]:     at InterceptingListener._callNext (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:568:42)
Oct 25 18:08:34 polisnode thunderhub[10086]:     at InterceptingListener.onReceiveStatus (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:618:8)
Oct 25 18:08:34 polisnode thunderhub[10086]:     at callback (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:847:24)
Oct 25 18:08:34 polisnode systemd[1]: thunderhub-system-mainnet.service: Main process exited, code=exited, status=1/FAILURE
Oct 25 18:08:34 polisnode systemd[1]: thunderhub-system-mainnet.service: Failed with result 'exit-code'.
Oct 25 18:08:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Service RestartSec=100ms expired, scheduling restart.
Oct 25 18:08:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Scheduled restart job, restart counter is at 4.
Oct 25 18:08:35 polisnode systemd[1]: Stopped Lightning Node Manager.
Oct 25 18:08:35 polisnode systemd[1]: Starting Lightning Node Manager...
Oct 25 18:08:35 polisnode systemd[1]: Started Lightning Node Manager.
Oct 25 18:08:36 polisnode thunderhub[12448]: ready - started server on http://localhost:4000
Oct 25 18:09:30 polisnode thunderhub[12448]: 2020-10-25 18:09:30 warn [THUB]: No SSO account available
Oct 25 18:10:15 polisnode thunderhub[12448]: 2020-10-25 18:10:15 warn [THUB]: No SSO account available
Oct 25 18:10:35 polisnode thunderhub[12448]: /usr/lib/thunderhub/node_modules/async/internal/onlyOnce.js:9
Oct 25 18:10:35 polisnode thunderhub[12448]:         if (fn === null) throw new Error("Callback was already called.");
Oct 25 18:10:35 polisnode thunderhub[12448]:                          ^
Oct 25 18:10:35 polisnode thunderhub[12448]: Error: Callback was already called.
Oct 25 18:10:35 polisnode thunderhub[12448]:     at /usr/lib/thunderhub/node_modules/async/internal/onlyOnce.js:9:32
Oct 25 18:10:35 polisnode thunderhub[12448]:     at args.lnd.(anonymous function).(anonymous function) (/usr/lib/thunderhub/node_modules/ln-service/node_modules/lightning/lnd_methods/info/get_node.js:101:20)
Oct 25 18:10:35 polisnode thunderhub[12448]:     at Object.onReceiveStatus (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:1212:9)
Oct 25 18:10:35 polisnode thunderhub[12448]:     at InterceptingListener._callNext (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:568:42)
Oct 25 18:10:35 polisnode thunderhub[12448]:     at InterceptingListener.onReceiveStatus (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:618:8)
Oct 25 18:10:35 polisnode thunderhub[12448]:     at callback (/usr/lib/thunderhub/node_modules/grpc/src/client_interceptors.js:847:24)
Oct 25 18:10:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Main process exited, code=exited, status=1/FAILURE
Oct 25 18:10:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Failed with result 'exit-code'.
Oct 25 18:10:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Service RestartSec=100ms expired, scheduling restart.
Oct 25 18:10:35 polisnode systemd[1]: thunderhub-system-mainnet.service: Scheduled restart job, restart counter is at 5.
Oct 25 18:10:35 polisnode systemd[1]: Stopped Lightning Node Manager.
Oct 25 18:10:35 polisnode systemd[1]: Starting Lightning Node Manager...
Oct 25 18:10:35 polisnode systemd[1]: Started Lightning Node Manager.
Oct 25 18:10:36 polisnode thunderhub[13002]: ready - started server on http://localhost:4000
Oct 25 18:10:45 polisnode thunderhub[13002]: 2020-10-25 18:10:45 error [THUB]: [ 400,
Oct 25 18:10:45 polisnode thunderhub[13002]:   'ExpectedAuthenticatedLndGrpcForGetInfoRequest',
Oct 25 18:10:45 polisnode thunderhub[13002]:   [length]: 2 ]
Oct 25 18:10:45 polisnode thunderhub[13002]: 2020-10-25 18:10:45 error [THUB]: [ 400,
Oct 25 18:10:45 polisnode thunderhub[13002]:   'ExpectedAuthenticatedLndToRetrieveChainBalance',
Oct 25 18:10:45 polisnode thunderhub[13002]:   [length]: 2 ]
Oct 25 18:10:45 polisnode thunderhub[13002]: 2020-10-25 18:10:45 error [THUB]: [ 400, 'ExpectedLndForPendingChainBalance', [length]: 2 ]
Oct 25 18:10:45 polisnode thunderhub[13002]: 2020-10-25 18:10:45 error [THUB]: [ 400, 'ExpectedLndGrpcApiForChannelBalanceQuery', [length]: 2 ]
Oct 25 18:10:51 polisnode thunderhub[13002]: 2020-10-25 18:10:51 warn [THUB]: No SSO account available
Oct 25 18:10:59 polisnode thunderhub[13002]: 2020-10-25 18:10:59 warn [THUB]: No SSO account available
```

</details>

**Additional context**
I've updated ThunderHub **and also** tried to experiment with the following nginx configuration:

```nginx
location /thunderhub {
        rewrite ^/thunderhub$ / break;
        rewrite ^/thunderhub(.*)$ $1 break;
        proxy_pass http://127.0.0.1:4000;
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_cookie_path ~^/thunderhub$ /;
        proxy_cookie_path /thunderhub/ /;
        proxy_cookie_path / /thunderhub;
}
```

Note the `proxy_cookie_path` at the end. The intent is to avoid sending the cookie to unrelated paths. It seems to work for receiving and other actions. If I comment it out, it still doesn't work but I wonder if it could be related somehow anyway.

While I doubt cookie path has anything to do with the problem, I'm mentioning it for completeness.

**Comment by apotdevin** [comment]
It was coming from the dependency update of `balanceofsatoshis`. Downgraded for now and should be fixed in version 0.9.17

**Comment by Kixunil** [comment]
Seems to work now, thanks!
